### ISSUE TITLE
Getting solution from cplex MIP models fails

### ISSUE THREAD

**Issue by cdiener** [opencobra/cobrapy#467]
Reduced costs can not be accessed in optlang cplex models without throwing an error. Might be preferable to fix that in optlang.

```Python
In [6]: from cobra.test import create_test_model

In [7]: mod = create_test_model("textbook")

In [8]: from cobra.flux_analysis import add_loopless

In [9]: mod.solver = "cplex"

In [10]: add_loopless(mod)

In [11]: sol = mod.optimize()
---------------------------------------------------------------------------
CplexSolverError                          Traceback (most recent call last)
<ipython-input-11-40ad0a3918b0> in <module>()
----> 1 sol = mod.optimize()

/home/cdiener/code/cobrapy/cobra/core/model.py in optimize(self, objective_sense, **kwargs)
    752                 objective_sense, original_direction)
    753         self.solver.optimize()
--> 754         solution = get_solution(self)
    755         self.objective.direction = original_direction
    756         return solution

/home/cdiener/code/cobrapy/cobra/core/solution.py in get_solution(model, reactions, metabolites)
    292     var_primals = model.solver.primal_values
    293     reduced = zeros(len(reactions))
--> 294     var_duals = model.solver.reduced_costs
    295     # reduced costs are not always defined, e.g. for integer problems
    296     if var_duals[rxn_index[0]] is None:

/home/cdiener/.local/lib/python3.5/site-packages/optlang/cplex_interface.py in reduced_costs(self)
    706     def reduced_costs(self):
    707         return collections.OrderedDict(
--> 708             zip((variable.name for variable in self.variables), self.problem.solution.get_reduced_costs()))
    709 
    710     @property

/home/cdiener/.local/lib/python3.5/site-packages/cplex/_internal/_subinterfaces.py in get_reduced_costs(self, *args)
   7608             return CPX_PROC.getdj(self._env._e, self._cplex._lp, a, b)
   7609         return apply_freeform_two_args(
-> 7610             getdj, self._cplex.variables.get_indices, args)
   7611 
   7612     def get_dual_values(self, *args):

/home/cdiener/.local/lib/python3.5/site-packages/cplex/_internal/_aux_functions.py in apply_freeform_two_args(fn, convert, args)
     71             raise TypeError("expecting name or index")
     72     elif len(args) == 0:
---> 73         return fn(0)
     74     else:
     75         raise WrongNumberOfArgumentsError()

/home/cdiener/.local/lib/python3.5/site-packages/cplex/_internal/_subinterfaces.py in getdj(a, b)
   7606         """
   7607         def getdj(a, b = self._cplex.variables.get_num() - 1):
-> 7608             return CPX_PROC.getdj(self._env._e, self._cplex._lp, a, b)
   7609         return apply_freeform_two_args(
   7610             getdj, self._cplex.variables.get_indices, args)

/home/cdiener/.local/lib/python3.5/site-packages/cplex/_internal/_procedural.py in getdj(env, lp, begin, end)
   1551     dj    = _safeDoubleArray(djlen)
   1552     status = CR.CPXXgetdj(env, lp, dj, begin, end)
-> 1553     check_status(env, status)
   1554     return LAU.array_to_list(dj, djlen)
   1555 

/home/cdiener/.local/lib/python3.5/site-packages/cplex/_internal/_procedural.py in __call__(self, env, status, from_cb)
    222                 else:
    223                     error_string = geterrorstring(env, status)
--> 224             raise CplexSolverError(error_string, env, status)
    225 
    226 check_status = StatusChecker()

CplexSolverError: CPLEX Error  1017: Not available for mixed-integer problems.
```

**Comment by Midnighter** [comment]
I meant to submit a PR to at least have a unified `optlang.SolverError` that we can catch in cobrapy. Still need to catch that there. Does that mean we have no MIP in our tests? Should have seen that before otherwise...

**Comment by cdiener** [comment]
There is one test (test_loopless_before) but it only uses glpk. Might want to add cplex as well if installed. I think optlang also lacks a method that tells you whether a problem is a MIP or not. One could add a hook in Variable addition in optlang.Model that sets an attribute.

**Comment by Midnighter** [comment]
A problem type flag could be nice. Always thought that this line was not very clean:
```Python
if var_duals[rxn_index[0]] is None:
```

**Comment by cdiener** [comment]
No, but the comment makes it clear. Just that it does not work always work with cplex apparently :smile:

**Comment by hredestig** [comment]
I also got stuck in this refactoring cameo, didn't make pr as not so pretty but could do for now until fixed with better errors from optlang.. #468

**Comment by hredestig** [comment]
Fixed for now in #468, new issue #469 to track the improvement
